x/provider/keeper/keeper_test.go

```diff
@@ -1,60 +1,178 @@
-	_, ok = s.keeper.Get(s.ctx, addr2)
-	s.Require().False(ok, "Get Provider failed")
-
-	s.T().Log("verify update provider")
-	host := "akash.domain.com"
-	msg.HostURI = host
-	s.keeper.Update(s.ctx, types.Provider(msg))
-	provider, _ := s.keeper.Get(s.ctx, ownerAddr)
-	s.Require().Equal(host, provider.HostURI, "Provider not updated")
-}
-
-func TestTestSuite(t *testing.T) {
-	suite.Run(t, new(TestSuite))
+func assertPanic(t *testing.T, f func()) {
+	defer func() {
+		if r := recover(); r == nil {
+			t.Errorf("the code did not panic")
+		}
+	}()
+	f()
+}
+
+func assertNoPanic(t *testing.T, f func()) {
+	defer func() {
+		if r := recover(); r != nil {
+			t.Errorf("the code must not panic")
+		}
+	}()
+	f()
+}
+
+func TestProviderCreate(t *testing.T) {
+	ctx, keeper := setupKeeper(t)
+	prov := testutil.Provider(t)
+
+	err := keeper.Create(ctx, prov)
+	require.NoError(t, err)
+
+	foundProv, found := keeper.Get(ctx, prov.Owner)
+	require.True(t, found)
+	require.Equal(t, prov, foundProv)
+}
+
+func TestProviderDuplicate(t *testing.T) {
+	ctx, keeper := setupKeeper(t)
+	prov := testutil.Provider(t)
+
+	err := keeper.Create(ctx, prov)
+	require.NoError(t, err)
+
+	err = keeper.Create(ctx, prov)
+	require.EqualError(t, err, types.ErrProviderExists.Error())
+}
+
+func TestProviderGetNonExisting(t *testing.T) {
+	ctx, keeper := setupKeeper(t)
+	prov := testutil.Provider(t)
+
+	foundProv, found := keeper.Get(ctx, prov.Owner)
+	require.False(t, found)
+	require.Equal(t, types.Provider{}, foundProv)
+}
+
+func TestProviderDeleteExisting(t *testing.T) {
+	ctx, keeper := setupKeeper(t)
+	prov := testutil.Provider(t)
+
+	err := keeper.Create(ctx, prov)
+	require.NoError(t, err)
+
+	assertPanic(t, func() {
+		keeper.Delete(ctx, prov.Owner)
+	})
+
+	foundProv, found := keeper.Get(ctx, prov.Owner)
+	require.True(t, found)
+	require.Equal(t, prov, foundProv)
+}
+
+func TestProviderUpdateNonExisting(t *testing.T) {
+	ctx, keeper := setupKeeper(t)
+	prov := testutil.Provider(t)
+
+	err := keeper.Update(ctx, prov)
+	require.EqualError(t, err, types.ErrProviderNotFound.Error())
+}
+
+func TestProviderUpdateExisting(t *testing.T) {
+	ctx, keeper := setupKeeper(t)
+	prov := testutil.Provider(t)
+
+	err := keeper.Create(ctx, prov)
+	require.NoError(t, err)
+
+	prov.HostURI = "akash.domain.com"
+	err = keeper.Update(ctx, prov)
+	require.NoError(t, err)
+
+	foundProv, found := keeper.Get(ctx, prov.Owner)
+	require.True(t, found)
+	require.Equal(t, prov, foundProv)
+}
+
+func TestWithProviders(t *testing.T) {
+	ctx, keeper := setupKeeper(t)
+	prov := testutil.Provider(t)
+	prov2 := testutil.Provider(t)
+
+	err := keeper.Create(ctx, prov)
+	require.NoError(t, err)
+
+	err = keeper.Create(ctx, prov2)
+	require.NoError(t, err)
+
+	count := 0
+
+	assertNoPanic(t, func() {
+		keeper.WithProviders(ctx, func(provider types.Provider) bool {
+			if !reflect.DeepEqual(provider, prov) && !reflect.DeepEqual(provider, prov2) {
+				panic("unknown provider")
```

Comment:
You can use `require.Fail(t,"unknown provider")`

Can also add a `Provider.Equals()` method if it makes sense to.